File: returned/python/today28/top_k_frequent_elements.py
# https://leetcode.com/problems/top-k-frequent-elements/description/
from typing import List
from collections import defaultdict

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = defaultdict(int)
        for n in nums:
            counter[n] += 1

        nums_sorted = sorted(counter.items(), key=lambda x: x[1], reverse=True)
        nums_sorted = [k for k,v in nums_sorted]
        return nums_sorted[:k]

File: csharp/Leet/Old/TopKFrequentElements.cs
// https://leetcode.com/problems/top-k-frequent-elements/description/
namespace Old.TopKFrequentElements
{
    public class Solution
    {
        public int[] TopKFrequent(int[] nums, int k)
        {
            var dict = new Dictionary<int, int>();
            foreach (var n in nums)
            {
                if (dict.TryGetValue(n, out var count))
                    dict[n] = count + 1;
                else
                    dict.Add(n, 1);
            }

            var buckets = new List<int>[nums.Length + 1];
            foreach (var (n, count) in dict)
            {
                if (buckets[count] == null) buckets[count] = new List<int>();
                buckets[count].Add(n);
            }

            var i = buckets.Length - 1;
            var result = new int[k];
            var r = 0;
            while (r < k)
            {
                if (buckets[i] != null && buckets[i].Count > 0)
                {
                    foreach (var key in buckets[i])
                    {
                        result[r] = key;
                        r++;
                        if (r >= k) break;
                    }
                }
                i--;
            }

            return result;
        }
    }
}
